### DISCUSSION TITLE
Usage of lagrange finite elements in parallel execution

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#16486] [Q&A General] [answered]
Dear all,
I wanted to use Lagrange family of finite elements for one my MOOSE parallel computation which requires certain nodal variables to be calculated along a specified boundary. I see that currently parallel computation is not possible with Lagrange family of elements and that the nodal based computation does strictly requires its presence. Here comes the error i received earlier.

__*** ERROR ***
The nodal patch recovery option, which calculates the Zienkiewicz-Zhu patch recovery for nodal variables (family = LAGRANGE), is not currently implemented for parallel runs. Run in serial if you must use the nodal patch capability__

My future work on MOOSE is performance based and so i cannot afford to run in a serial mode. For your reference i am attaching here the input file i am using to run the problem. I would welcome any suggestions/idea on how to accomplish a parallel computation with Lagrange elements and nodal patch capabilities.
Kind regards,
Arun
[PieceWise_test.txt](https://github.com/idaholab/moose/files/5689262/PieceWise_test.txt)

**Comment by jiangwen84** [comment]
First, nodal patch recovery has not been implemented for parallel yetl, and I am not sure whether it will be done very soon.

Secondly, do you really need or want to use nodal patch recovery? Looks like you want to evaluate stress on a specific boundary. For this purpose, I suggest you consider evaluate and/or integrate stress using the quadratures on the boundary. That would be most accurate way.

**Comment by abarun22** [comment]
Hi Wen,
Thanks for the response. Well, my intention is to extract the mean
stresses/strains from a selected boundary, irrespective of how it is
computed i.e., through nodal or element based. I see that
'ElementAverageValue' does not include a boundary identifier to facilitate
this sort of selective calculation. Broadly speaking i am interested in
creating a so called 'path' engulfing the boundary which is subjected to
necking and measure the stresses and strains (as the loading progresses)
along this reducing area. I would be glad to hear from you as how i can
output these data, possibly with the quadrature rules you mentioned, so
that i can plot the true stress/strain results .
Kind regards,
Arun

**Reply by jiangwen84** [reply]
I would compute those materials on the boundary, and you can find some boundary material tests under moose/test/tests/materials/boundary_material

But you need to implement something like `ElementIntegralMaterialProperty`, but for side (boundary), not block. You can follow    `SideIntegralVariablePostprocessor` to implement that.

**Reply by abarun22** [reply]
Hi Wen,
Thanks for the response. Indeed my problem of stress dis-agreement at large strains has been solved by the inclusion of 'volumetric locking correction' in to the numerical scheme of things and so i may not be in an urgent need to implement  'ElementIntegralMaterialProperty' to be able to apply along the boundaries for stress calculation. However another advantage of using a nodal patch recovery is to avoid the rendering problems associated with the element based calculations. The attached file shows the difference between the images rendered by peacock for a strain contour done with 'ElementAverageValue' and 'AverageNodalVariableValue'. The later one provide dual benefits of selective calculation of field variables and perfect rendering of pretty pictures, but with a penalty in the form of sequential run and the associated slowness. The problem which takes just over half an hour might take 10-15 days to compute the whole solution, which is never going to be a preferred option. I would like to know if there are any other methods to obtain a smooth contour without having to use the time consuming nodal methods?
Kind regards,
Arun
![rendering_issues](https://user-images.githubusercontent.com/40111479/102916433-77697c00-447b-11eb-99b1-21703fa57d85.png)

**Reply by GregVernon** [reply]
If you're just concerned with rendering pretty pictures, you could use ParaView's filter: `Cell Data to Point Data` :

![image](https://user-images.githubusercontent.com/868479/103030919-04fba900-451a-11eb-80dc-a057dffe3469.png)

**Comment by jiangwen84** [comment]
In addition to @GregVernon 's solution, you can also use FIRST order monomial for element variable. MOOSE/libmesh uses least square fitting to get the value at the nodes.

[stress_xx]
    order = FIRST
    type = MONOMIAL
[]

**Comment by abarun22** [comment]
Thanks for the suggestions. I dont really get how we go from element
variables (as Wen described) to the nodal results. Are there any options to
be included in the input file to enforce this conversion?

**Reply by jiangwen84** [reply]
Let us use  modules/tensor_mechanics/test/tests/ad_linear_elasticity/applied_strain.i  as example

Add following blocks to the input file. This will output stress_xx as a nodal variable.

[AuxVariables]
  [stress_xx]
    order = FIRST
    family = MONOMIAL
  []
[]

[AuxKernels]
  [stress_xx]
    type = ADRankTwoAux
    variable = stress_xx
    rank_two_tensor = stress
    index_i = 0
    index_j = 0
  []
[]

`TensorMechancsMasterAction` `generate_output` option will automatically use Elemental variable. So if you want to output any strain or stress as a nodal variable, you have to manually add the AuxVariable and AuxKernels like what is shown above.

**Comment by hugary1995** [comment, marked as answer]
I think what you really want is something similar to the SidesetReaction postprocessor available in the tensor_mechanics module.

**Comment by abarun22** [comment]
Thanks for your suggestion wen, i could now be able to get smooth
stress/strain contours after conversion to nodal variable output. Thinking
on the lines of Gary, could i ask if it is possible to extend
'ElementAverageValue' to accept domain boundaries, so that just by
specifying this identifier, we could calculate the results specifically to
a part of the model. I am not pretty sure if this can be done practically,
but just a thought to ponder upon.

I am happy to work on the SidesetReaction postprocessor if that's something
we could develop quickly.

**Reply by jiangwen84** [reply]
The `ElementAverageValue` iderives from `ElementIntegralPostprocessor` which loops all the elements. So you need to implement a new one that derives from `SideIntegralPostprocessor` which loops over a side/boundary.
